File: packages/typescript/src/features/advanced-capabilities/providers/contacts.ts
import { requireProviderSpec } from "../../../generated/spec-helpers.ts";
import type { RelationshipsService } from "../../../services/relationships.ts";
import type {
	IAgentRuntime,
	Memory,
	Provider,
	ProviderResult,
	State,
} from "../../../types/index.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("CONTACTS");

export const contactsProvider: Provider = {
	name: spec.name,
	description: spec.description,
	get: async (
		runtime: IAgentRuntime,
		_message: Memory,
		_state: State,
	): Promise<ProviderResult> => {
		const relationshipsService = runtime.getService(
			"relationships",
		) as RelationshipsService;
		if (!relationshipsService) {
			runtime.logger.warn(
				"[ContactsProvider] RelationshipsService not available",
			);
			return { text: "", values: {}, data: {} };
		}

		// Get all contacts
		const contacts = await relationshipsService.searchContacts({});

		if (contacts.length === 0) {
			return {
				text: "No contacts in relationships.",
				values: { contactCount: 0 },
				data: {},
			};
		}

		// Get entity details and categorize
		const contactDetails = await Promise.all(
			contacts.map(async (contact) => {
				const entity = await runtime.getEntityById(contact.entityId);
				const displayName =
					typeof contact.customFields.displayName === "string"
						? contact.customFields.displayName
						: null;
				return {
					id: contact.entityId,
					name: entity?.names[0] || displayName || "Unknown",
					categories: contact.categories,
					tags: contact.tags,
					preferences: contact.preferences,
					lastModified: contact.lastModified,
				};
			}),
		);

		// Group by category
		const grouped: Record<string, typeof contactDetails> = {};
		for (const contact of contactDetails) {
			for (const cat of contact.categories) {
				const bucket = grouped[cat];
				if (bucket) {
					bucket.push(contact);
				} else {
					grouped[cat] = [contact];
				}
			}
		}

		const lines: string[] = [];
		lines.push(`You have ${contacts.length} contacts in your relationships:`);

		const categoryCounts: Record<string, number> = {};
		for (const category in grouped) {
			const items = grouped[category];
			if (!items) continue;
			categoryCounts[category] = items.length;
			lines.push(
				"",
				`${category.charAt(0).toUpperCase() + category.slice(1)}s (${items.length}):`,
			);
			for (const item of items) {
				let line = `- ${item.name}`;
				if (item.tags.length > 0) {
					line += ` [${item.tags.join(", ")}]`;
				}
				lines.push(line);
			}
		}

		const textSummary = lines.join("\n").trim();

		return {
			text: textSummary,
			values: {
				contactCount: contacts.length,
				...categoryCounts,
			},
			data: categoryCounts,
		};
	},
};

File: packages/python/elizaos/features/basic_capabilities/providers/contacts.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("CONTACTS")


async def get_contacts_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    from elizaos.features.advanced_capabilities.services.relationships import RelationshipsService

    relationships_service = runtime.get_service("relationships")
    if not relationships_service or not isinstance(relationships_service, RelationshipsService):
        return ProviderResult(text="", values={}, data={})

    contacts = await relationships_service.get_all_contacts()

    if not contacts:
        return ProviderResult(
            text="No contacts in relationships.", values={"contactCount": 0}, data={}
        )

    contact_details: list[dict[str, str]] = []
    for contact in contacts:
        entity = await runtime.get_entity(str(contact.entity_id))
        name = entity.name if entity and entity.name else "Unknown"
        contact_details.append(
            {
                "id": str(contact.entity_id),
                "name": name,
                "categories": ",".join(contact.categories),
                "tags": ",".join(contact.tags),
            }
        )

    grouped: dict[str, list[dict[str, str]]] = {}
    for detail in contact_details:
        for cat in detail["categories"].split(","):
            cat = cat.strip()
            if cat:
                grouped.setdefault(cat, []).append(detail)

    text_summary = f"You have {len(contacts)} contacts in your relationships:\n"

    for category, items in grouped.items():
        text_summary += f"\n{category.capitalize()}s ({len(items)}):\n"
        for item in items:
            text_summary += f"- {item['name']}"
            if item["tags"]:
                text_summary += f" [{item['tags']}]"
            text_summary += "\n"

    category_counts = {cat: len(items) for cat, items in grouped.items()}

    return ProviderResult(
        text=text_summary.strip(),
        values={
            "contactCount": len(contacts),
            **category_counts,
        },
        data=category_counts,
    )


contacts_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_contacts_context,
    dynamic=_spec.get("dynamic", True),
    position=_spec.get("position"),
)
